cpp/src/parquet/arrow/writer.cc

```diff
@@ -518,6 +518,36 @@ class FileWriterImpl : public FileWriter {
     return Status::OK();
   }
 
+  Status WriteRecordBatch(const ::arrow::RecordBatch& batch) override {
+    auto schema = batch.schema();
+    auto size = batch.num_rows();
+
+    if (!schema->Equals(*schema_, false)) {
+      return Status::Invalid("RecordBatch schema does not match this writer's. batch:'",
+                             schema->ToString(), "' this:'", schema_->ToString(), "'");
+    }
+
+    RETURN_NOT_OK(NewRowGroup(size));
+    for (int i = 0; i < batch.num_columns(); i++) {
+      RETURN_NOT_OK(WriteColumnChunk(*batch.column(i)));
+    }
+
+    return Status::OK();
+  }
+
+  Status WriteRecordBatchReader(::arrow::RecordBatchReader* reader) override {
+    auto schema = reader->schema();
+
+    if (!schema->Equals(*schema_, false)) {
```

Comment:
The schema check is performed on each invocation of `WriteRecordBatch`. Perhaps that's the safest thing